### DISCUSSION TITLE
Autocomplete inside a Datagrid

### DISCUSSION THREAD

**Discussion by zawarq** [Megabit/Blazorise#3824] [General]
Hi

I have placed an autocomplete as the edit template inside a datagrid. For context, I have a Patient model for which I am trying to specify the City. Here is the autocomplete markup:

`<DataGridColumn Field="@nameof(Patient.CityId)" Caption="City" Editable Filterable="false">
        <DisplayTemplate>
             <Span>@context.CityName</Span>
         </DisplayTemplate>
         <EditTemplate>
             <Autocomplete TItem="City"
              TValue="string"
              Data="@cities"
              TextField="@((City x) => x.Name)"
              ValueField="@((City x) => x.Id.ToString())"
              SelectedValue="@(context.CellValue.ToString())"
              SelectedValueChanged="@((value) => SelectedCityChanged(context, value))"
              Placeholder="Search..."
              Filter="AutocompleteFilter.StartsWith"
              FreeTyping
              CustomFilter="@(( item, searchValue ) => item.Name.IndexOf( searchValue, 0, StringComparison.CurrentCultureIgnoreCase ) >= 0 )">
              <NotFoundContent Context="autoContext"> Sorry... @autoContext was not found </NotFoundContent>
              <ItemContent Context="autoContext">
                <Div Flex="Flex.InlineFlex.JustifyContent.Between" Width="Width.Is100">
                    <Div Margin="Margin.Is2.FromBottom">@autoContext.Text</Div>
                    <Small>@autoContext.Item.StateName, @autoContext.Item.CountryName</Small>
                </Div>
              </ItemContent>
            </Autocomplete>
        </EditTemplate>
    </DataGridColumn>`

The SelectedValueChanged event handler is as follows:

`private void SelectedCityChanged(CellEditContext<Patient> context, string value)
    {
        if (value == null)
            value = string.Empty;

        context.CellValue = value;
    }`

Once I Save the changes in my datagrid, I expect the following methods to execute correctly:

`void OnRowUpdating (CancellableRowChange<Patient, Dictionary<string, object>> c)
    {
        validateInsertOrUpdate(c);
    }
    async Task OnRowUpdated (SavedRowItem<Patient, Dictionary<string, object>> c)
    {
        await AlertService.ProcessMessage(await PatientService.Update(c.Item), c.Item, CrudOperationType.Updated);
        StateHasChanged();
    }`

However, I get the following error after OnRowUpdating successfully executes, but before OnRowUpdated is called:

`Microsoft.AspNetCore.Components.WebAssembly.Rendering.WebAssemblyRenderer[100]
      Unhandled exception rendering component: Specified cast is not valid.
System.InvalidCastException: Specified cast is not valid.
   at Blazorise.DataGrid.DataGridColumn`1[[NutritionProViewModels.Practice.Patient, NutritionProViewModels, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null]].SetValue(Patient item, Object value)
   at Blazorise.DataGrid.DataGrid`1.<Save>d__43[[NutritionProViewModels.Practice.Patient, NutritionProViewModels, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null]].MoveNext()
   at Blazorise.DataGrid._BaseDataGridRowEdit`1.<SaveWithValidation>d__9[[NutritionProViewModels.Practice.Patient, NutritionProViewModels, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null]].MoveNext()
   at Microsoft.AspNetCore.Components.ComponentBase.CallStateHasChangedOnAsyncCompletion(Task task)
   at Blazorise.Button.ClickHandler()
   at Microsoft.AspNetCore.Components.ComponentBase.CallStateHasChangedOnAsyncCompletion(Task task)
   at Microsoft.AspNetCore.Components.RenderTree.Renderer.GetErrorHandledTask(Task taskToHandle, ComponentState owningComponentState)`

Can you tell me what I am doing wrong? Thanks!

**Comment by zawarq** [comment]
I fixed this by changing my SelectedCityChanged implementation to the following:

`private void SelectedCityChanged(CellEditContext<Patient> context, string value)
    {
        if (value == null)
            value = "0";

        context.CellValue = Convert.ToInt32(value);
    }`

By casting the value before assigning to CellValue, it fixes the model conversion problem that I was facing. IMO a numeric string ought to work, so worth investigating by the dev team :)

**Reply by David-Moreira** [reply]
It's just the way that it's built at the moment. No generic support, so users should make sure they assign the `context.CellValue` a value with the same type as the column type.

If I remember correctly there were technical reasons for this.
Maybe we can reinvestigate @stsrki ? Another idea would be to add some helpers that do these boxing/conversion for you, like `context.AssignValue`

But yes, that's how you should do it at the moment, make sure you assign/box the exact same type as it is configured on your `DataGridColumn`.

**Reply by stsrki** [reply]
If I remember correctly, there was no way to match the TValue of the Items inside of the rows, with the columns and rows needed to pass to the DataGrid internals. I've tried several approaches but none would work or compile even remotely.
